src/opam.ml

```diff
@@ -131,15 +131,18 @@ end = struct
     let open Promise.Syntax in
     let path = Switch.path opam switch in
     let switch_state_filepath = Opam_path.switch_state path |> Path.to_string in
-    let* file_exists = Fs.exists switch_state_filepath in
-    if file_exists then
-      let+ file_content = Fs.readFile switch_state_filepath in
-      let { OpamParserTypes.file_contents; _ } =
-        OpamParser.FullPos.string file_content switch_state_filepath
-        |> OpamParser.FullPos.to_opamfile
-      in
-      Some file_contents
-    else Promise.return None
+    Promise.then_
+      (Fs.exists switch_state_filepath)
+      ~fulfilled:(fun file_exists ->
+        if file_exists then
+          let+ file_content = Fs.readFile switch_state_filepath in
+          let { OpamParserTypes.file_contents; _ } =
+            OpamParser.FullPos.string file_content switch_state_filepath
+            |> OpamParser.FullPos.to_opamfile
+          in
+          Some file_contents
+        else Promise.return None)
+      ~rejected:(fun x -> Promise.reject x)
```

Comment:
This code is effectively the same as previous and susceptible to a race condition. When you first check if the file exists and _then_ try reading it, the file that you checked to exist may already have been removed. 

I'll push a fix :-)

Reply:
> This code is effectively the same as previous and susceptible to a race condition. When you first check if the file exists and _then_ try reading it, the file that you checked to exist may already have been removed.
> 
> I'll push a fix :-)

Oh right, I misunderstood. But I'm curious to see the solution.